### ISSUE TITLE
question about the DYNAMIC VISUAL TOKENIZER

### ISSUE THREAD

**Issue by bpwl0121** [jy0205/LaVIT#10]
hi, 

thanks for your awesome work, I have a question regarding the loss func in DYNAMIC VISUAL TOKENIZER.
the codebook generate the discrete token Vq, how do you train your DYNAMIC VISUAL TOKENIZER by the cosine similarity and the Mask loss since part of it is discrete 

thx

**Comment by jy0205** [comment]
Thaks for your attention! We use the Gumbel-Softmax trick to maintain the gradient passing.

**Comment by jy0205** [comment]
Thanks for your attention~ We use the Gumbel-Softmax trick to maintain the gradient passing during training.

**Comment by bpwl0121** [comment]
> Thanks for your attention~ We use the Gumbel-Softmax trick to maintain the gradient passing during training.

Gumbel-Softmax trick does its job for sure, but it only relate to the Mask loss, I think. how is the codebook trained?  cause you do not use the loss like vq-vae, the common loss for codebook training
![image](https://github.com/jy0205/LaVIT/assets/38959389/312233b2-9880-4335-b901-ee352d3b9dd8)

**Comment by jy0205** [comment]
Sorry for the late reply~ In our implementation, we first train the token predictor and freeze other modules (without feature quantize), which aims to train a good predictor to select the most informative patches. Then, we freeze the token predictor and conduct the joint VQ training with the objective of Eq.4 to update other modules in the paper.

**Comment by bpwl0121** [comment]
> Sorry for the late reply~ In our implementation, we first train the token predictor and freeze other modules (without feature quantize), which aims to train a good predictor to select the most informative patches. Then, we freeze the token predictor and conduct the joint VQ training with the objective of Eq.4 to update other modules in the paper.

ah, I see
thanks for your reply 👍
